File: bench/wasm/driver.mjs
import createModule from './bench.mjs';
import {makeViews, workCopy, memoryNames, statNames, stepNames, jsonReplacer} from './views.mjs';
export {jsonReplacer};
export const fixtureNames = Object.freeze(['pyramid','rain','piles','chains','churn','table','inverted']);
export const timestep = Math.fround(1/60);
const snake = name => name.replace(/[A-Z]/g, c => `_${c.toLowerCase()}`);
const round = value => Number(value.toPrecision(9)); // Native report %.9g compatibility.
const qualityNames = ['penetration_max','cached_penetration_max','translation_drift_max','rotation_drift_max',
  'linear_speed_max','angular_speed_max','joint_error_max','cached_support_force_mean','supported_weight'];
const settingNames = ['linear_drag','angular_drag','linear_speed_max','contact_hertz','contact_damping_ratio',
  'contact_push_velocity_max','restitution_threshold','joint_hertz','joint_damping_ratio',
  'sleep_speed_max','sleep_angular_speed_max','sleep_time_min'];
function uint(value, low, high, name) {
  if (!Number.isInteger(value) || value < low || value > high) throw new RangeError(`Invalid ${name}`);
  return value;
}
function wordsToObject(names, values, offset = 0) { return Object.fromEntries(names.map((name,i)=>[name,values[offset+i]])); }
const work = values => Object.fromEntries(Object.entries(workCopy(values)).map(([name,value])=>[snake(name),value]));
export async function createBenchmarkModule({memoryBytes = 64*1024*1024, wasmUrl} = {}) {
  uint(memoryBytes,2*1024*1024,512*1024*1024,'memoryBytes');
  if (memoryBytes % 65536) throw new RangeError('Memory budget must be 64 KiB aligned');
  if (wasmUrl !== undefined && typeof wasmUrl !== 'string' && !(wasmUrl instanceof URL)) throw new TypeError('Invalid wasmUrl');
  let m;
  try { m = await createModule({wasmMemory:new WebAssembly.Memory({initial:memoryBytes/65536,maximum:memoryBytes/65536}),
    ...(wasmUrl === undefined ? {} : {locateFile:()=>String(wasmUrl)})}); }
  catch(cause) { throw new Error('Benchmark module initialization failed; check asset URL and fixed memory budget',{cause}); }
  const worlds = new Set(); let disposed = false;
  const live = () => { if (disposed) throw new Error('Benchmark module is disposed'); };
  return Object.freeze({
    get memory() { live(); return {linear_memory:memoryBytes,stack:1048576,static_end:m._sl_wasm_static_end(),
      heap_base:m._sl_wasm_heap_base(),allocator_used:m._sl_wasm_allocator_used()}; },
    create(fixture, {sleep = false, warmup = 120, steps = 600} = {}) {
      live(); const id = fixtureNames.indexOf(fixture);
      if (id < 0 || typeof sleep !== 'boolean') throw new TypeError('Invalid fixture or sleep policy');
      uint(warmup,0,10000,'warmup'); uint(steps,1,10000,'steps');
      let ptr = m._sl_wasm_bench_create(id,sleep?1:0,warmup,steps);
      if (!ptr) return null;
      try {
        const adapter = m._sl_wasm_bench_adapter(ptr), buffer = m.HEAPU32.buffer;
        const io = {ptr:adapter, of:new Float32Array(buffer,m._sl_wasm_output_f32(adapter),64),
          ou:new Uint32Array(buffer,m._sl_wasm_output_u32(adapter),32)};
        const status = new Uint32Array(buffer,m._sl_wasm_bench_words(ptr),13);
        const quality = new Float64Array(buffer,m._sl_wasm_bench_quality(ptr),25);
        const counters = new BigUint64Array(buffer,m._sl_wasm_bench_counters(ptr),2);
        m._sl_wasm_bench_report(ptr);
        const config = {bodyCapacity:status[8],contactCapacity:status[9],jointCapacity:status[10]};
        const views = makeViews(m,io,config,()=>{throw new Error('Benchmark queries do not issue mutable world handles');});
        const valid = () => { live(); if (!ptr) throw new Error('Benchmark is disposed'); };
        const invalidate = () => { views.state.valid = false; views.queryState.valid = false; views.rayState.valid = false; };
        const invoke = fn => { valid(); invalidate(); return !!m[fn](ptr); };
        const value = Object.freeze({
          fixture, warmup, steps,
          prepare:()=>invoke('_sl_wasm_bench_prepare'), mutate:()=>invoke('_sl_wasm_bench_mutate'),
          step:()=>invoke('_sl_wasm_bench_step'), sample:()=>invoke('_sl_wasm_bench_sample'),
          get snapshot() { valid(); return views.snapshot; },
          refreshSnapshot(diagnostics = false) {
            valid(); if (typeof diagnostics !== 'boolean') throw new TypeError('Invalid diagnostics flag');
            if (views.state.revision === Number.MAX_SAFE_INTEGER) throw new RangeError('Snapshot revision exhausted');
            invalidate();
            if (!m._sl_wasm_snapshot_refresh(adapter,diagnostics?1:0)) return false;
            views.refreshCopy(io.ou[3],io.ou[1],io.ou[2]);
            const state = views.state;
            state.bodyCount=io.ou[0]; state.contactCount=io.ou[1]; state.jointCount=io.ou[2];
            state.geometryUpdates=io.ou[3]; ++state.revision; state.valid=true;
            return true;
          },
          report() {
            valid(); m._sl_wasm_bench_report(ptr);
            m._sl_wasm_stats_read(adapter);
            const stats=wordsToObject(statNames,io.ou), last=wordsToObject(stepNames,io.ou,13);
            const lastWork=work(views.work), cumulative=work(views.work.subarray(13));
            m._sl_wasm_memory_read(adapter);
            const memory=Object.fromEntries(memoryNames.map((name,i)=>[snake(name.replace(/Bytes$/,'')),io.ou[i]]));
            const settings={fixture_version:1,seed:status[1],warmup_steps:warmup,measured_steps:steps,
              dt_seconds:round(timestep),substeps:status[11],body_capacity:config.bodyCapacity,
              contact_capacity:config.contactCapacity,joint_capacity:config.jointCapacity,sleep_enabled:sleep,
              gravity:[round(quality[9]),round(quality[10])],friction:round(quality[23]),restitution:round(quality[24]),
              ...Object.fromEntries(settingNames.map((name,i)=>[name,round(quality[11+i])]))};
            return {scene:fixture,settings,semantic_digest:counters[0].toString(16).padStart(16,'0'),drops:counters[1],
              counts:{awake_dynamics:stats.awakeDynamicCount,sleeping_dynamics:stats.sleepingDynamicCount,
                bodies:stats.bodyCount,contacts:stats.contactCount,joints:stats.jointCount,pairs:stats.pairCount,
                pair_capacity:stats.pairCapacity,body_high:stats.bodyCountHigh,contact_high:stats.contactCountHigh,joint_high:stats.jointCountHigh},
              step:{dynamic_bodies:last.dynamicBodyCount,kinematic_bodies:last.kinematicBodyCount,
                contact_constraints:last.contactConstraintCount,joint_constraints:last.jointConstraintCount,
                substeps:last.substepCount,islands:last.islandCount,island_bodies_max:last.islandBodyCountMax,
                islands_executed:last.islandExecutedCount,islands_skipped:last.islandSkippedCount,work:lastWork},
              cumulative_work:cumulative,memory_bytes:memory,
              quality:{window_steps:status[7],...Object.fromEntries(qualityNames.map((name,i)=>[name,round(quality[i])]))},
              driver:{index:status[4],phase:status[5],failed:!!status[6],complete:status[4]===warmup+steps,
                context_bytes:m._sl_wasm_bench_bytes(),adapter_bytes:m._sl_wasm_adapter_bytes(adapter),
                output_bytes:views.outputBytes,context_base_bytes:m._sl_wasm_context_bytes()}};
          },
          dispose() { if (ptr) { invalidate(); m._sl_wasm_bench_destroy(ptr); ptr=0; worlds.delete(value); } }
        });
        worlds.add(value); return value;
      } catch (error) { m._sl_wasm_bench_destroy(ptr); if (error instanceof RangeError) return null; throw error; }
    },
    dispose() { if (!disposed) { for (const world of worlds) world.dispose(); disposed=true; } }
  });
}

File: bench/wasm/driver.c
#include "driver.h"
#include "../fixtures.h"
#include "../quality.h"
#include "context.h"
#include <stdlib.h>

enum { PHASE_READY, PHASE_PREPARED, PHASE_MUTATED, PHASE_STEPPED };
struct sl_wasm_bench {
    sl_wasm_context adapter;
    sl_bench_scene scene;
    sl_bench_quality quality;
    uint64_t drops;
    uint32_t fixture, warmup, steps, index, phase;
    bool failed;
    uint32_t words[13];
    double values[25];
    uint64_t counters[2];
};
sl_wasm_bench *sl_wasm_bench_create(uint32_t fixture, uint32_t sleep_enabled,
                                    uint32_t warmup, uint32_t steps)
{
    if (fixture >= SL_BENCH_FIXTURE_COUNT || sleep_enabled > 1u ||
        warmup > SL_BENCH_STEP_COUNT_MAX || steps == 0u ||
        steps > SL_BENCH_STEP_COUNT_MAX) {
        return NULL;
    }
    sl_wasm_bench *b = calloc(1u, sizeof(*b));
    if (b == NULL) {
        return NULL;
    }
    if (!sl_bench_scene_build(&b->scene, &b->adapter.world, fixture,
                              sleep_enabled != 0u)) {
        free(b);
        return NULL;
    }
    b->adapter.config = b->scene.config;
    b->adapter.config.contact_capacity =
        sl_world_contact_capacity(&b->adapter.world);
    if (!sl_wasm_storage_init(&b->adapter)) {
        sl_wasm_world_dispose(&b->adapter);
        free(b);
        return NULL;
    }
    b->fixture = fixture;
    b->warmup = warmup;
    b->steps = steps;
    return b;
}
void sl_wasm_bench_destroy(sl_wasm_bench *b)
{
    if (b != NULL) {
        sl_wasm_world_dispose(&b->adapter);
        free(b);
    }
}
sl_wasm_context *sl_wasm_bench_adapter(sl_wasm_bench *b)
{
    return b != NULL ? &b->adapter : NULL;
}
size_t sl_wasm_bench_bytes(void)
{
    return sizeof(sl_wasm_bench);
}
static bool phase_valid(const sl_wasm_bench *b, uint32_t phase)
{
    return b != NULL && !b->failed && b->index < b->warmup + b->steps &&
           b->phase == phase;
}
static uint32_t window_start(const sl_wasm_bench *b)
{
    const uint32_t window =
        b->steps < SL_BENCH_QUALITY_WINDOW ? b->steps : SL_BENCH_QUALITY_WINDOW;
    return b->warmup + b->steps - window;
}
bool sl_wasm_bench_prepare(sl_wasm_bench *b)
{
    if (!phase_valid(b, PHASE_READY)) {
        return false;
    }
    if (b->index == window_start(b)) {
        sl_bench_reference_capture(&b->scene);
    }
    b->phase = PHASE_PREPARED;
    return true;
}
bool sl_wasm_bench_mutate(sl_wasm_bench *b)
{
    if (!phase_valid(b, PHASE_PREPARED)) {
        return false;
    }
    if (!sl_bench_scene_mutate(&b->scene, b->index)) {
        b->failed = true;
        return false;
    }
    b->phase = PHASE_MUTATED;
    return true;
}
bool sl_wasm_bench_step(sl_wasm_bench *b)
{
    if (!phase_valid(b, PHASE_MUTATED)) {
        return false;
    }
    sl_world_step(&b->adapter.world, SL_BENCH_TIMESTEP);
    b->phase = PHASE_STEPPED;
    return true;
}
bool sl_wasm_bench_sample(sl_wasm_bench *b)
{
    if (!phase_valid(b, PHASE_STEPPED)) {
        return false;
    }
    b->drops += sl_world_contact_drop_count(&b->adapter.world);
    if (b->index >= b->warmup &&
        !sl_bench_quality_sample(&b->scene, &b->quality,
                                 b->index >= window_start(b))) {
        b->failed = true;
        return false;
    }
    ++b->index;
    b->phase = PHASE_READY;
    return true;
}
bool sl_wasm_bench_report(sl_wasm_bench *b)
{
    if (b == NULL) {
        return false;
    }
    const sl_bench_quality *q = &b->quality;
    const sl_vec2 gravity = sl_world_get_gravity(&b->adapter.world);
    const double values[] = {
        (double)q->penetration_max,
        (double)q->cached_penetration_max,
        (double)q->translation_drift_max,
        (double)q->rotation_drift_max,
        (double)q->linear_speed_max,
        (double)q->angular_speed_max,
        (double)q->joint_error_max,
        q->window_steps != 0u ? q->support_force_sum / (double)q->window_steps
                              : 0.0,
        q->supported_weight,
        (double)gravity.x,
        (double)gravity.y,
        (double)sl_world_get_linear_drag(&b->adapter.world),
        (double)sl_world_get_angular_drag(&b->adapter.world),
        (double)sl_world_get_linear_speed_max(&b->adapter.world),
        (double)SL_CONTACT_HERTZ_DEFAULT,
        (double)SL_CONTACT_DAMPING_RATIO_DEFAULT,
        (double)SL_CONTACT_PUSH_VELOCITY_MAX_DEFAULT,
        (double)SL_RESTITUTION_THRESHOLD_DEFAULT,
        (double)SL_JOINT_HERTZ_DEFAULT,
        (double)SL_JOINT_DAMPING_RATIO_DEFAULT,
        (double)SL_SLEEP_SPEED_MAX_DEFAULT,
        (double)SL_SLEEP_ANGULAR_SPEED_MAX_DEFAULT,
        (double)SL_SLEEP_TIME_MIN_DEFAULT,
        b->fixture == 1u ? (double)0.4f : (double)0.6f,
        b->fixture == 1u ? (double)0.1f : 0.0
    };
    for (uint32_t i = 0u; i < 25u; ++i) {
        b->values[i] = values[i];
    }
    const uint32_t words[] = { b->fixture,
                               b->scene.seed,
                               b->warmup,
                               b->steps,
                               b->index,
                               b->phase,
                               b->failed ? 1u : 0u,
                               q->window_steps,
                               b->adapter.config.body_capacity,
                               b->adapter.config.contact_capacity,
                               b->adapter.config.joint_capacity,
                               sl_world_get_substep_count(&b->adapter.world),
                               b->scene.config.sleep_enabled ? 1u : 0u };
    for (uint32_t i = 0u; i < 13u; ++i) {
        b->words[i] = words[i];
    }
    b->counters[0] = sl_bench_digest(&b->adapter.world, b->scene.bodies,
                                     b->scene.body_count);
    b->counters[1] = b->drops;
    return true;
}
const uint32_t *sl_wasm_bench_words(const sl_wasm_bench *b)
{
    return b->words;
}
const double *sl_wasm_bench_quality(const sl_wasm_bench *b)
{
    return b->values;
}
const uint64_t *sl_wasm_bench_counters(const sl_wasm_bench *b)
{
    return b->counters;
}
